pkg/controller/servicediscovery/servicediscovery_controller.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ServiceDiscovery
+	err = c.Watch(&source.Kind{Type: &submarinerv1alpha1.ServiceDiscovery{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource Pods and requeue the owner ServiceDiscovery
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &submarinerv1alpha1.ServiceDiscovery{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileServiceDiscovery implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileServiceDiscovery{}
+
+// ReconcileServiceDiscovery reconciles a ServiceDiscovery object
+type ReconcileServiceDiscovery struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ServiceDiscovery object and makes changes based on the state read
+// and what is in the ServiceDiscovery.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileServiceDiscovery) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ServiceDiscovery")
+
+	// Fetch the ServiceDiscovery instance
+	instance := &submarinerv1alpha1.ServiceDiscovery{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			daemonSet := &appsv1.DaemonSet{}
+			opts := []DeleteAllOfOption{
+				InNamespace(request.NamespacedName.Namespace),
+				MatchingLabels{"app": appName},
+			}
+			err := r.client.DeleteAllOf(context.TODO(), daemonSet, opts...)
+			return reconcile.Result{}, err
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Create Lighthouse-agent
+	lighthouseAgent := newLighthouseAgent(instance)
+
+	// Set ServiceDiscovery instance as the owner and controller
+	if err := controllerutil.SetControllerReference(instance, lighthouseAgent, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if this lighthouse-agent already exists
+	found := &corev1.Pod{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: lighthouseAgent.Name, Namespace: lighthouseAgent.Namespace}, found)
```

Comment:
Shouldn't this check if the `DaemonSet` already exists and create-or-update, ie reuse `reconcileDaemonSet`?

Reply:
Done.